text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,96 @@
+- `const char *protocols`: protocols separated with a semi-colon (ex: `"RTMP;RTMPS"`), required to register the output.
+- `const char *protocols_prefixes`: URL protocols prefixes separated with a semi-colon `"rtmp://;rtmps://"`, optional.
+  - Only one prefix per protocol in `protocols`.
+  - Those prefixes should be in the same order as `protocols`.
+  - Meant to be used for protocol auto-detection.
+  - While registering if present, the number of prefix shall be less or equal to the number of protocols.
+- Two attributes already existent will be required if non-'codec agnostic':
+  - `const char *encoded_video_codecs`: video codecs supported by the protocol separated with a semi-colon, required if `OBS_OUTPUT_VCODEC_AGNOSTIC` is not set.
+  - `const char *encoded_audio_codecs`: audio codecs supported by the protocol separated with a semi-colon, required if `OBS_OUTPUT_ACODEC_AGNOSTIC` is not set.
+
+Asside the getter for protocol attribute, adding to the API those functions:
+- `bool obs_output_is_protocol_registered(const char *protocol)`: return true if an output with the protocol is registered.
+- `const char *obs_output_get_prefix_protocol(const char *prefix)`: return the protocol bound to the prefix.
+- `bool obs_enum_output_service_types(const char *protocol, size_t idx, const char **id)`: enumerate all outputs types compatible with the given protocol.
+- `bool obs_enum_output_protocols_prefixes(size_t idx, const char **prefix)`: enumerate all registered URL prefixes.
+- `bool obs_enum_output_protocols(size_t idx, const char **prefix)`: enumerate all registered protocol.
+
+### About protocols
+RTMP and RTMPS will on be considered compatible only with H264 and AAC.
+
+HLS will have no prefix set and not "auto-detectable".
+
+HLS, SRT and RIST are codec agnostic, since they use MPEG/TS container.
+
+FTL will be considered compatible only with H264 and opus because of the deprecation of the protocol in OBS Studio.
+
+## Services API
+Since a streaming service may not accept all the codec usable by a protocol, adding a way to set supported codecs is required.
+
+Adding to `obs_service_info`:
+- `const char *(*get_protocols)(void *data)`: return the protocol used by the service. It will allow multi-protocol services in the future.
+- `const char **(*get_supported_video_codecs)(void *data)`: video codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+- `const char **(*get_supported_audio_codecs)(void *data)`: audio codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+
+### About `rtmp-services`
+
+This plugin will:
+- for `rtmp_common` only accept H264, HEVC for Youtube HLS, AAC and opus for FTL protocol.
+- for `rtmp_custom` it will depends on the auto-detected protocol.
+
+## UI
+
+### Audio encoders
+If "Custom..." service is selected, the UI should allow to choose between compatible audio encoders. Just a combobox to choose which defaults to AAC.
```

Comment:
Allowing other codecs than aac (eyeing opus ...) should in parallel be done with better support in Settings>output>audio since atm only aac bitrates are enumerated.

Reply:
Maybe I use too few words too describe adding the possibility to choose between AAC and opus when the protocol (and the service) is compatible with both.

Reply:
I put this one as resolved, feel free to review the new version of this paragraph.

Reply:
>  better support in Settings>output>audio since atm only aac bitrates are enumerated

I get what you mean now, and yes better support will be (is) implemented.